### ISSUE TITLE
azurerm_key_vault_access_policy Doesn't Support All Valid Key_Permissions

### ISSUE THREAD

**Issue by JStLouisFsv** [hashicorp/terraform-provider-azurerm#17866]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.0.6

### AzureRM Provider Version

3.16.0

### Affected Resource(s)/Data Source(s)

azurerm_key_vault_access_policy

### Terraform Configuration Files

```hcl
Provider does not allow for total management of Key permissions.

Current provider allowed permissions are: "Backup", "Create", "Decrypt", "Delete", "Encrypt", "Get", "Import", "List",<br /> "Purge", "Recover", "Restore", "Sign", "UnwrapKey", "Update", "Verify", "WrapKey"

Provider update should also include permissions: "Release", "Rotate", "GetRotationPolicy", "SetRotationPolicy"
```

### Debug Output/Panic Output

```shell
{"@level":"error","@message":"Error: expected key_permissions.16 to be one of [Backup Create Decrypt Delete Encrypt Get Import List Purge Recover Restore Sign UnwrapKey Update Verify WrapKey], got Release","@module":"terraform.ui","@timestamp":"2022-08-03T15:57:12.073564Z","diagnostic":{"severity":"error","summary":"expected key_permissions.16 to be one of [Backup Create Decrypt Delete Encrypt Get Import List Purge Recover Restore Sign UnwrapKey Update Verify WrapKey], got Release","detail":"","address":"module.key-vault.azurerm_key_vault_access_policy.sp_pols","range":{"filename":".terraform/modules/key-vault/main.tf","start":{"line":100,"column":29,"byte":2811},"end":{"line":100,"column":50,"byte":2832}},"snippet":{"context":"resource \"azurerm_key_vault_access_policy\" \"sp_pols\"","code":"  key_permissions         = local.kvkeyperms_full","start_line":100,"highlight_start_offset":28,"highlight_end_offset":49,"values":[]}},"type":"diagnostic"}
{"@level":"error","@message":"Error: expected key_permissions.17 to be one of [Backup Create Decrypt Delete Encrypt Get Import List Purge Recover Restore Sign UnwrapKey Update Verify WrapKey], got Rotate","@module":"terraform.ui","@timestamp":"2022-08-03T15:57:12.076817Z","diagnostic":{"severity":"error","summary":"expected key_permissions.17 to be one of [Backup Create Decrypt Delete Encrypt Get Import List Purge Recover Restore Sign UnwrapKey Update Verify WrapKey], got Rotate","detail":"","address":"module.key-vault.azurerm_key_vault_access_policy.sp_pols","range":{"filename":".terraform/modules/key-vault/main.tf","start":{"line":100,"column":29,"byte":2811},"end":{"line":100,"column":50,"byte":2832}},"snippet":{"context":"resource \"azurerm_key_vault_access_policy\" \"sp_pols\"","code":"  key_permissions         = local.kvkeyperms_full","start_line":100,"highlight_start_offset":28,"highlight_end_offset":49,"values":[]}},"type":"diagnostic"}
{"@level":"error","@message":"Error: expected key_permissions.18 to be one of [Backup Create Decrypt Delete Encrypt Get Import List Purge Recover Restore Sign UnwrapKey Update Verify WrapKey], got GetRotationPolicy","@module":"terraform.ui","@timestamp":"2022-08-03T15:57:12.082746Z","diagnostic":{"severity":"error","summary":"expected key_permissions.18 to be one of [Backup Create Decrypt Delete Encrypt Get Import List Purge Recover Restore Sign UnwrapKey Update Verify WrapKey], got GetRotationPolicy","detail":"","address":"module.key-vault.azurerm_key_vault_access_policy.sp_pols","range":{"filename":".terraform/modules/key-vault/main.tf","start":{"line":100,"column":29,"byte":2811},"end":{"line":100,"column":50,"byte":2832}},"snippet":{"context":"resource \"azurerm_key_vault_access_policy\" \"sp_pols\"","code":"  key_permissions         = local.kvkeyperms_full","start_line":100,"highlight_start_offset":28,"highlight_end_offset":49,"values":[]}},"type":"diagnostic"}
{"@level":"error","@message":"Error: expected key_permissions.19 to be one of [Backup Create Decrypt Delete Encrypt Get Import List Purge Recover Restore Sign UnwrapKey Update Verify WrapKey], got SetRotationPolicy","@module":"terraform.ui","@timestamp":"2022-08-03T15:57:12.085270Z","diagnostic":{"severity":"error","summary":"expected key_permissions.19 to be one of [Backup Create Decrypt Delete Encrypt Get Import List Purge Recover Restore Sign UnwrapKey Update Verify WrapKey], got SetRotationPolicy","detail":"","address":"module.key-vault.azurerm_key_vault_access_policy.sp_pols","range":{"filename":".terraform/modules/key-vault/main.tf","start":{"line":100,"column":29,"byte":2811},"end":{"line":100,"column":50,"byte":2832}},"snippet":{"context":"resource \"azurerm_key_vault_access_policy\" \"sp_pols\"","code":"  key_permissions         = local.kvkeyperms_full","start_line":100,"highlight_start_offset":28,"highlight_end_offset":49,"values":[]}},"type":"diagnostic"}
```

### Expected Behaviour

The valid permissions should have been assigned to the access policy

### Actual Behaviour

The TFE plan failed because the provider is not allowing the required permissions

### Steps to Reproduce

Attempt to apply a policy which includes any one of the missing valid permissions: "Release", "Rotate", "GetRotationPolicy", "SetRotationPolicy"

### Important Factoids

None

### References

Not that I am aware of

**Comment by mikemadeja** [comment]
Do you have an example of the Terraform code you are trying to apply to re-create the issue?

**Comment by JStLouisFsv** [comment]
Within the module we created, in the main.tf file we are calling the resource from the provider like this to grant full permissions to the subscription service principal:

# -
# - Key Vault Access Policy for Service Principal
# -
resource "azurerm_key_vault_access_policy" "sp_pols" {
  key_vault_id            = azurerm_key_vault.kv.id
  tenant_id               = data.azurerm_client_config.current.tenant_id
  object_id               = data.azurerm_client_config.current.object_id
  certificate_permissions = local.kvcertperms_full
  key_permissions         = local.kvkeyperms_full
  secret_permissions      = local.kvsecretperms_full
  storage_permissions     = local.kvstorperms_full

  depends_on = [
    azurerm_key_vault.kv,
  ]
}

And in the locals.tf file, we have the full permissions defined as follows:

  # -
  # - Access Policy Permissions (Full) 
  # -
  kvkeyperms_full = [
    "Get", "List", "Update", "Create", "Import", "Delete", "Recover", "Backup", "Restore",
    "Decrypt", "Encrypt", "UnwrapKey", "WrapKey", "Verify", "Sign", "Purge",
    "Release", "Rotate", "GetRotationPolicy", "SetRotationPolicy"
  ]

  kvsecretperms_full = [
    "Backup", "Delete", "Get", "List", "Purge", "Recover", "Restore", "Set"
  ]

  kvcertperms_full = [
    "Backup", "Create", "Delete", "DeleteIssuers", "Get", "GetIssuers", "Import", "List", "ListIssuers",
    "ManageContacts", "ManageIssuers", "Purge", "Recover", "Restore", "SetIssuers", "Update"
  ]

  kvstorperms_full = [
    "Backup", "Delete", "DeleteSAS", "Get", "GetSAS", "List", "ListSAS", "Purge",
    "Recover", "RegenerateKey", "Restore", "Set", "SetSAS", "Update"
  ]

If I remove the "Release", "Rotate", "GetRotationPolicy", "SetRotationPolicy" permissions from kvkeyperms_full variable, everything runs fine.  However that means you cannot fully manage the Azure Key Vault Access Policy, and manual intervention is required to grant the additional permissions.

"Release", "Rotate", "GetRotationPolicy", "SetRotationPolicy" are valid permissions from the Azure CLI.  I think the simple resolution to this is to simply add these permissions to the provider as valid values.

**Comment by mikemadeja** [comment]
I attempted to fix the issue by added the properties you listed and was able to do it and test each one. I submitted a PR, hopefully I got everything!

**Comment by JordanKergoat** [comment]
Thanks @mikemadeja your PR will hopefully me released in 3.19 : [Pull request](https://github.com/hashicorp/terraform-provider-azurerm/pull/17895)

Milestone : [3.19](https://github.com/hashicorp/terraform-provider-azurerm/milestone/194)

**Comment by aristosvo** [comment]
Closed as feature is available!
